source/binding/OperatorExpressions.cpp

```diff
@@ -1596,10 +1676,52 @@ ConstantValue StreamingConcatenationExpression::evalImpl(EvalContext& context) c
     std::vector<ConstantValue> values;
     values.reserve(streams().size());
     for (auto stream : streams()) {
-        ConstantValue v = stream->eval(context);
-        if (!v)
+        auto cv = stream->operand->eval(context);
+        if (!cv)
             return nullptr;
-        values.emplace_back(std::move(v));
+        if (stream->with) {
+            auto& arrayType = *stream->operand->type;
+            auto range = Bitstream::evaluateWith(arrayType, *stream->with, context);
+            if (!range)
+                return nullptr;
+            ASSERT(range->lower() >= 0);
+
+            // If the range expression evaluates to a range greater than the extent of the array
+            // size, the entire array is streamed, and the remaining items are generated using the
+            // nonexistent array entry value
+            const auto emptyValue = arrayType.getArrayElementType()->getDefaultValue();
+            if (range->left == range->right) {
+                if (size_t(range->left) >= cv.size())
+                    cv = emptyValue;
+                else
+                    cv = std::move(cv).at(size_t(range->left));
+            }
+            else if (range->lower() > 0 || range->width() != cv.size()) {
+                auto upper = static_cast<uint32_t>(range->upper());
+                auto lower = static_cast<uint32_t>(range->lower());
+                auto size = static_cast<uint32_t>(cv.size());
+                auto more = upper >= size ? upper - size + 1 : 0;
+                upper = std::min(upper + 1, size);
+                if (cv.isUnpacked()) {
```

Comment:
This code seems duplicated in Bitstream.cpp, can it be shared?

Reply:
Yes, I consolidate them into a new helper function declared in Bitstream.h:
```c++
    /// Resize a constant array value from [0:size-1] to [lower:upper]
    static ConstantValue resizeToRange(const Type& arrayType, ConstantValue&& value,
                                       ConstantRange range, bool keepArray = false);
```